Automattic/newspack-blocks
# feat: support subscriptions gifting

### All Submissions:






### Changes proposed in this Pull Request:



Adds support for the [WooCommerce Subscriptions Gifting extension](https://github.com/woocommerce/woocommerce-subscriptions-gifting) to our modal checkout flow.

Because of limitations in how the WCSG extension handles its own form fields and validation, this PR implements custom form fields for inputting gift recipient info during checkout and validating that input via WCSG's own validation methods.

### How to test the changes in this Pull Request:

1. Install and activate the [WooCommerce Subscriptions Gifting extension](https://github.com/woocommerce/woocommerce-subscriptions-gifting)
2. Check out this branch and start a new subscription transaction (donation or non-donation via Checkout Button).
3. Confirm that the gift subscription options are available in the first modal screen:

<img width="651" alt="Screenshot 2024-01-04 at 12 39 06 PM" src="https://github.com/Automattic/newspack-blocks/assets/2230142/bca99307-dcb1-4c42-8de9-7ef38b783f4e">

4. Confirm that checking the "This donation/transaction is a gift" box toggles the recipient email field.
5. UNCHECK the box and confirm that you can complete the transaction without entering anything into the email field.
6. Start a new transaction and CHECK the box. Confirm that you can't proceed if you don't enter an email address or if you enter a non-email address string into the gift recipient field.

<img width="654" alt="Screenshot 2024-01-04 at 12 39 22 PM" src="https://github.com/Automattic/newspack-blocks/assets/2230142/6329690c-a5fc-4379-a0a0-5f5fb4dc73d4">

7. Confirm that you can't proceed if you enter the same email address as the billing email address into the gift recipient field.

<img width="643" alt="Screenshot 2024-01-04 at 12 39 31 PM" src="https://github.com/Automattic/newspack-blocks/assets/2230142/b96db157-3125-47c5-929e-13d3a72b7835">

8. Enter a valid email address into the gift recipient field and confirm that you see it in the billing details summary on the next modal screen:

<img width="673" alt="Screenshot 2024-01-04 at 12 42 42 PM" src="https://github.com/Automattic/newspack-blocks/assets/2230142/d0748927-396e-417e-a7a6-598732c1c4ba">

9. Complete the transaction and confirm that the subscription is listed with both buyer and recipient info in WP admin, and that the recipient receives an email address informing them of their new gift subscription.

<img width="310" alt="Screenshot 2024-01-04 at 12 54 40 PM" src="https://github.com/Automattic/newspack-blocks/assets/2230142/4e6d2887-58b0-4365-8b61-0722d15b24e3">
<img width="625" alt="Screenshot 2024-01-04 at 12 54 15 PM" src="https://github.com/Automattic/newspack-blocks/assets/2230142/e71ba2a2-5db1-4596-96c4-47adaead1b2f">

10. Smoke test a transaction through the regular Shop/Cart/Checkout pages and confirm that the gift options work there too (they should be untouched from the base Subscriptions Gifting plugin experience).

### Other information:
